src/util/configInheritance.tsx

```diff
@@ -0,0 +1,246 @@
+import { InstanceAndProfileFormValues } from "components/forms/instanceAndProfileFormValues";
+import { LxdProfile } from "types/profile";
+import { CreateInstanceFormValues } from "pages/instances/CreateInstance";
+import { EditInstanceFormValues } from "pages/instances/EditInstance";
+import { isDiskDevice, isNicDevice } from "util/devices";
+import { LxdDiskDevice, LxdNicDevice } from "types/device";
+import { ProjectFormValues } from "pages/projects/CreateProject";
+import { ConfigurationRowFormikValues } from "components/ConfigurationRow";
+import { ConfigField } from "types/config";
+import { useQuery } from "@tanstack/react-query";
+import { queryKeys } from "util/queryKeys";
+import { fetchConfigOptions } from "api/server";
+import { toConfigFields } from "util/config";
+import { getInstanceKey } from "util/instanceConfigFields";
+import { useParams } from "react-router-dom";
+import { fetchProfiles } from "api/profiles";
+import { getProjectKey } from "util/projectConfigFields";
+import { StorageVolumeFormValues } from "pages/storage/forms/StorageVolumeForm";
+import { fetchStoragePool } from "api/storage-pools";
+import { getStorageVolumeDefault, getVolumeKey } from "util/storageVolume";
+import { getNetworkDefault } from "util/networks";
+
+export interface ConfigRowMetadata {
+  value?: string;
+  source: string;
+  configField?: ConfigField;
+}
+
+export const getConfigRowMetadata = (
+  values: ConfigurationRowFormikValues,
+  name: string,
+): ConfigRowMetadata => {
+  switch (values.type) {
+    case "instance":
+    case "profile":
+      return getInstanceRowMetadata(values, name);
+    case "project":
+      return getProjectRowMetadata(values, name);
+    case "storageVolume":
+      return getStorageVolumeRowMetadata(values, name);
+    case "network":
+      return getNetworkRowMetadata(name);
+  }
+};
+
+const getInstanceRowMetadata = (
+  values: InstanceAndProfileFormValues | ProjectFormValues,
+  name: string,
+): ConfigRowMetadata => {
+  const { data: configOptions } = useQuery({
+    queryKey: [queryKeys.configOptions],
+    queryFn: fetchConfigOptions,
+  });
+
+  const configFields = toConfigFields(configOptions?.configs.instance ?? {});
+  const configKey = getInstanceKey(name);
+  const configField = configFields.find((item) => item.key === configKey);
+
+  const { project } = useParams<{ project: string }>();
+  const { data: profiles = [] } = useQuery({
+    queryKey: [queryKeys.profiles],
+    queryFn: () => fetchProfiles(project ?? ""),
+    enabled: Boolean(project),
+  });
+
+  // inherited values from applied profiles
+  if (Object.prototype.hasOwnProperty.call(values, "profiles")) {
+    const appliedProfiles = [
+      ...(values as CreateInstanceFormValues | EditInstanceFormValues).profiles,
+    ].reverse();
+    for (const profileName of appliedProfiles) {
+      const profile = profiles.find((profile) => profile.name === profileName);
```

Comment:
nit: This is potentially an O^2 operation. We could iterate profiles before, construct a lookup of profile names? Although there might not be that many profiles so probably a non-issue

Reply:
I see quite a few functions needs to search through profiles, maybe having a function that generates a profileName lookup makes sense?

Reply:
An instance won't have more than 1-5 profiles max. But still this code is quite redundant and can be optimized, as well as reused in all places to make the functions simpler.